porch/controllers/downstreampackages/pkg/controllers/downstreampackage/downstreampackage_controller.go

```diff
@@ -0,0 +1,366 @@
+	dp *api.DownstreamPackage,
+	upstream *porchapi.PackageRevision) (*porchapi.PackageRevision, error) {
+	// First, check if a downstream package exists. If not, just return nil. The
+	// caller will create one.
+	downstream := r.getDownstreamPR(dp)
+	if downstream == nil {
+		return nil, nil
+	}
+
+	// Determine if the downstream package needs to be updated. If not, return
+	// the downstream package as-is.
+	if r.isUpToDate(dp, downstream) {
+		return downstream, nil
+	}
+
+	if downstream.Spec.Lifecycle == porchapi.PackageRevisionLifecyclePublished {
+		var err error
+		downstream, err = r.copyPublished(ctx, downstream, dp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return r.updateDraft(ctx, downstream, upstream)
+}
+
+func (r *DownstreamPackageReconciler) getDownstreamPR(dp *api.DownstreamPackage) *porchapi.PackageRevision {
+	downstream := dp.Spec.Downstream
+
+	var latestPublished *porchapi.PackageRevision
+	// the first package revision number that porch assigns is "v1",
+	// so use v0 as a placeholder for comparison
+	latestVersion := "v0"
+
+	for _, pr := range r.packageRevs.Items {
+		// look for the downstream package in the target repo
+		if pr.Spec.RepositoryName != downstream.Repo ||
+			pr.Spec.PackageName != downstream.Package {
+			continue
+		}
+		// check that the downstream package was created by this DownstreamPackage object
+		owned := false
+
+		// TODO: Implement the "AdoptionPolicy" field, which allows the user to decide if
+		//  the controller should adopt existing package revisions or ignore them. For now,
+		//  we just ignore them.
+		for _, owner := range pr.OwnerReferences {
+			if owner.UID == dp.UID {
+				owned = true
+			}
+		}
+		if !owned {
+			// this downstream package doesn't belong to us
+			continue
+		}
+
+		// Check if this PR is a draft. We should only have one draft created by this controller at a time,
+		// so we can just return it.
+		if pr.Spec.Lifecycle != porchapi.PackageRevisionLifecyclePublished {
+			return &pr
+		} else {
+			latestPublished, latestVersion = compare(&pr, latestPublished, latestVersion)
+		}
+	}
+
+	return latestPublished
+}
+
+func compare(pr, latestPublished *porchapi.PackageRevision, latestVersion string) (*porchapi.PackageRevision, string) {
+	switch cmp := semver.Compare(pr.Spec.Revision, latestVersion); {
+	case cmp == 0:
+		// Same revision.
+	case cmp < 0:
+		// current < latest; no change
+	case cmp > 0:
+		// current > latest; update latest
+		latestVersion = pr.Spec.Revision
+		latestPublished = pr.DeepCopy()
+	}
+	return latestPublished, latestVersion
+}
+
+// determine if the downstream PR needs to be updated
+func (r *DownstreamPackageReconciler) isUpToDate(dp *api.DownstreamPackage, downstream *porchapi.PackageRevision) bool {
+	upstreamLock := downstream.Status.UpstreamLock
+	lastIndex := strings.LastIndex(upstreamLock.Git.Ref, "/")
+	if strings.HasPrefix(upstreamLock.Git.Ref, "drafts") {
```

Comment:
Can't we just check the `lifecycle` field on the PackageRevision for this?

Reply:
We have the target upstream package revision (the one we want to update to), and the current downstream package revision. What we do not have is the upstream package that the current downstream package is _currently_ based on - which is what we need to check if the downstream package revision is up to date.

All we have is the reference to that current upstream package revision in the downstream package's `upstreamLock.Git.Ref` - so no, we don't have a PackageRevision.Lifecycle field to refer to here.

Fetching that current upstream package revision involves tedious loops and string parsing of `upstreamLock.Git.Ref` (and somewhere in that logic would include checking for the `drafts` prefix so we can parse the ref properly), so IMO just checking for the `drafts` prefix is the best thing to do here.